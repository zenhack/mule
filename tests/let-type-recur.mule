#XFAIL: recursive let type
let type int-list
  = Cons { hd : int, tl : int-list }
  | Nil {}
in
(Cons { hd = 4, tl = Nil {} } : int-list)
