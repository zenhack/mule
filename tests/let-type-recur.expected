
Desugared:

	(letrec
	 (types
	  ((int-list
	    (rec int-list
	     (union
	      ((Cons (record (types ()) (values ((tl int-list) (hd int)))))
	       (Nil (record (types ()) (values ())))))))))
	 (values ())
	 (:
	  ((fn x (Cons x))
	   (letrec (types ()) (values ((hd (Integer 4)) (tl ((fn x (Nil x)) {}))))
	    (.= hd hd (.= tl tl {}))))
	  int-list))

inferred type:

	(rec a
	 (union
	  ((Cons (record (types ()) (values ((tl a) (hd int)))))
	   (Nil (record (types ()) (values ()))))))

Runtime term:

	(App
	 ((Lam (0 () (Ctor (Cons (Var 0)))))
	  (LetRec
	   (((0 (Const (Integer 4)))
	     (0 (App ((Lam (0 () (Ctor (Nil (Var 0))))) (Record ())))))
	    (Update (old (Update (old (Record ())) (label tl) (field (Var 1))))
	     (label hd) (field (Var 0)))))))

Evaluated:

	(Ctor
	 (Cons (Record ((hd (Const (Integer 4))) (tl (Ctor (Nil (Record ()))))))))
